src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,102 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] elementData = new Object[DEFAULT_CAPACITY];
+    private int size;
+
     @Override
     public void add(T value) {
-
+        if (size == elementData.length) {
+            elementData = grow();
+        }
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size + 1) {
+            throw new ArrayListIndexOutOfBoundsException("Index is out of bounds");
+        }
+        if (size == elementData.length || index > size) {
+            elementData = grow();
+        }
+        Object[] elementsAfterIndex = new Object[this.size - index + 1];
+        for (int i = 0; i < elementsAfterIndex.length; i++) {
+            elementsAfterIndex[i] = elementData[index + i];
+        }
+        elementData[index] = value;
+        for (int i = index + 1; i <= size; i++) {
+            elementData[i] = elementsAfterIndex[i - index - 1];
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            T elem = (T) list.get(i);
+            add(elem);
+        }
+    }
 
+    private Object[] grow() {
+        int oldCapacity = elementData.length;
+        int newCapacity = oldCapacity + (oldCapacity >> 1);
+        return elementData = Arrays.copyOf(elementData, newCapacity);
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index > size - 1) {
+            throw new ArrayListIndexOutOfBoundsException("Index is out of bounds");
+        }
+        return (T) elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index < 0 || index > size - 1) {
```

Comment:
```suggestion
        if (index < 0 || index >= size) {
```
extract this check to a private utility method